proxy/ptunnel/commands/tunnel.go

```diff
@@ -182,42 +271,65 @@ func IsBackgroundFork() bool {
 //    as file descriptor number 3 (this is part of the ExtraFiles API in cmd.Exec).
 //    This hopefully will get us back in this function. A simple fork() would have been significantly
 //    simpler, but the behavior with threads is undefined and evil.
-func (r *Tunnel) OpenAndListen(proxy *url.URL, id, host string, port uint16, cookie *http.Cookie, hostport string) error {
-	var listener net.Listener
+func (r *Tunnel) OpenAndListen(ctx context.Context, proxy *url.URL, id, host string, port uint16, cookie *http.Cookie, network string, localAddr string) error {
+	var listener knetwork.FileListener
 	if r.Background && IsBackgroundFork() {
 		var err error
-		listener, err = net.FileListener(os.NewFile(3, "listener"))
+		fl, err := net.FileListener(os.NewFile(3, "listener"))
 		if err != nil {
 			return err
 		}
+		// Shouldn't ever fail, based on the implementation of `net.fileListener()`:
+		// https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/net/file_unix.go;drc=d7a3fa120db1f8ab9e02ea8fccd0cc8699bf9382;l=89
+		// and the fact that the listener created below only supports protocols
+		// whose corresponding net.Listener concrete implementations are also
+		// knetwork.FileListeners.
+		listener = fl.(knetwork.FileListener)
+
+		// If using a UNIX domain socket - make sure the socket path is cleaned
+		// up on close.
+		if network == "unix" && localAddr != "" {
+			listener = &knetwork.CleanupListener{FileListener: listener, Path: localAddr}
+		}
 	} else {
 		err := r.CanConnect(proxy, id, host, port, cookie)
 		if err != nil {
 			return err
 		}
 
-		addr, err := net.ResolveTCPAddr("tcp", hostport)
-		if err != nil {
-			return err
+		if localAddr == "" {
+			return fmt.Errorf("no local listen address set")
 		}
 
-		tlisten, err := net.ListenTCP("tcp", addr)
+		l, err := net.Listen(network, localAddr)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to listen on %s %s: %w", network, localAddr, err)
+		}
+
+		var ok bool
+		listener, ok = l.(knetwork.FileListener)
+		if !ok {
+			return fmt.Errorf("got net.Listener with no File() method: %T", l)
 		}
 		if r.Background {
-			err := r.RunBackground(tlisten)
-			tlisten.Close()
+			err := r.RunBackground(listener)
+			// If using a UNIX domain socket, calling Close() here unlinks it
+			// from the FS. The forked process will still be able to listen
+			// since it has an open FD, but no other process can dial it, which
+			// kinda defeats the purpose.
+			//
+			// Instead, let the forked process clean up when it is signalled.
```

Comment:
should this just delete the file anyway? RunBackground failed, it's probably "correct" that we no longer use that failed background process? and allow another enkit to bind on the same tunnel?

(the forked child will eventually die, but at least other children can now start?)

Reply:
(the most annoying part about UDS is that if the program that created one crashes/dies, the file is left behind, and trying to bind on it now returns EADDRINUSE - so it's generally better to remove/close aggressively than risk leaving files behind)

Reply:
(except that when the background process dies, it may remove the file descriptor another process created ... ugh. Internet wisdom uses flock, but seems like overkill here - use your best judgement after testing)

Reply:
Clarified the error when using a socket and getting EADDRINUSE. I tried to make the cleanup automatic via some mechanism with flock, but this was nontrivial and makes this code even more branchy. Manual cleanup should be fine for now to reduce the risk of adding nastier bugs (deadlocks, etc.)

Reply:
Also added a Close() if RunBackground fails (but not if it succeeds, as per the reasoning in the comment)